onnxruntime/core/providers/cpu/tensor/concat.cc

```diff
@@ -17,59 +17,82 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx, int input_count, Prep
 
+  // cache num of elements in tensor for later use
+  // as it's expensive to call Size() on TensorShape over and over
+  std::vector<size_t> tensor_num_elements(input_count);
   // Ensure all of the non concatenated axes match each other
   for (int index = 1; index < input_count; index++) {
+    size_t num_elements = 1;
     tensor_pointer = ctx->Input<Tensor>(index);
     if (tensor_pointer == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
-    auto& data_n = *tensor_pointer;
-    // Ensure all the other axes match
-    auto dimension_count = inputs_0.Shape().NumDimensions();
-    for (int axis_index = 0; axis_index < dimension_count; axis_index++) {
+    auto& inputs_n = *tensor_pointer;
+    const auto& inputs_n_dims = inputs_n.Shape().GetDims();
+    const size_t inputs_n_rank = inputs_n_dims.size();
+    ORT_ENFORCE(inputs_n_rank == inputs_0_rank, "Ranks of input data are different, cannot concatenate them");
+    // Ensure all the other (non-concat) axes match
+    for (int axis_index = 0; axis_index < inputs_0_rank; axis_index++) {
+      num_elements *= inputs_n_dims[axis_index];
       if (axis_index == axis)
         continue;
-      ORT_RETURN_IF_NOT(data_n.Shape()[axis_index] == inputs_0.Shape()[axis_index], "Non concat axis dimensions must match: Axis ", axis_index, " has mismatched dimensions of ", data_n.Shape()[axis_index], " and ", inputs_0.Shape()[axis_index]);
+      ORT_RETURN_IF_NOT(inputs_n_dims[axis_index] == inputs_0_dims[axis_index],
+                        "Non concat axis dimensions must match: Axis ", 
+                        axis_index, " has mismatched dimensions of ", inputs_n_dims[axis_index], 
+                        " and ", inputs_0_dims[axis_index]);
     }
+    tensor_num_elements[index] = num_elements;
   }
 
   // Calculate the size of the concatenated axis, and verify all other dimensions match
   size_t concat_axis_size = 0;
   for (int index = 0; index < input_count; index++) {
     tensor_pointer = ctx->Input<Tensor>(index);
-    if (tensor_pointer == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
     concat_axis_size += tensor_pointer->Shape()[int(axis)];
   }
 
   // Calculate the shape of the output tensor
-  std::vector<int64_t> dims;
-  for (int dimension_index = 0; dimension_index < inputs_0.Shape().NumDimensions(); dimension_index++)
-    dims.emplace_back(inputs_0.Shape()[dimension_index]);
+  std::vector<int64_t> dims(inputs_0_rank);
+  size_t num_elements = 1; // cache size of the first input along the way
+  for (int dimension_index = 0; dimension_index < inputs_0_rank; dimension_index++) {
+    dims[dimension_index] = inputs_0_dims[dimension_index];
+    num_elements *= inputs_0_dims[dimension_index];
+  }
+  tensor_num_elements[0] = num_elements;
   dims[axis] = concat_axis_size;
   TensorShape outputShape(dims);
+ 
+  auto& concat_result = *ctx->Output(0, outputShape);
```

Comment:
name outputShape doesn't follow conventions.

Reply:
Again - just stuck to the style of the earlier implementation - will fix